dbms/include/DB/Interpreters/Context.h

```diff
@@ -276,8 +276,9 @@ class Context
 	  */
 	void resetCaches() const;
 
-	const Cluster & getCluster(const std::string & cluster_name) const;
-	std::shared_ptr<Clusters> getClusters() const;
+	Clusters & getClusters() const;
+	std::shared_ptr<Cluster> getCluster(const std::string & cluster_name, bool throw_on_error = true) const;
+	void setClustersConfig(ConfigurationPtr config);
```

Comment:
We not just apply passed config, we also store config to apply it in the future (on first getClusters() call). So we need to store it (at least for clusters).
Currently, to do this we just copy shared_ptr (Poco::AutoPtr).
If we pass reference, we will must make copy of Poco::Util::AbstractConfiguration created inside ConfigReloader.